### PR TITLE
docs(agents): clamp is not a portable gradient guard

### LINKED ISSUE #4229: torch.clamp's boundary gradient changed between 2.9.1 and 2.14.0, so kornia's clamp-as-gradient-guard pattern silently fails on the older half of its supported range
## Summary

`torch.clamp`'s gradient **at the clamp bound itself** changed between torch 2.9.1 and 2.14.0: it used to pass
the incoming gradient through (`1.0`), and now returns `0.0`. kornia declares `torch>=2.5.1` and its PR CI runs
**three** torch versions (2.5.1 and 2.14.0 in the main matrix, 2.9.1 on the half-precision legs), so the two ends
of the supported range disagree.

That matters because kornia repeatedly uses `clamp` as a *gradient* guard directly in front of an operator whose
own derivative is unbounded exactly at the clamp bound — `clamp(min=0.0).sqrt()`, `clamp(-1, 1).acos()`,
`clamp(-1, 1).asin()`. Wherever the bound **is** the singular point, the guard silently works on 2.14 and does
**not** work on 2.5.1 / 2.9.1: the clamp hands the unbounded derivative straight through and the result is `inf`
or `nan`.

Measured, same machine, CPU:

| | torch 2.5.1 | torch 2.9.1 | torch 2.14.0 |
| --- | --- | --- | --- |
| `clamp(x, min=0.0)` grad at `x = 0` | `1.0` | `1.0` | **`0.0`** |
| `clamp(w, -1, 1)` grad at `w = 1` | `1.0` | `1.0` | **`0.0`** |
| `clamp(x, min=0.0).sqrt()` grad at `x = 0` | `inf` | `inf` | `0.0` |
| `clamp(w, -1, 1).acos()` grad at `w = 1` | `-inf` | `-inf` | `0.0` |
| bare `acos` grad at `w = 1` | `-inf` | `-inf` | `-inf` |
| bare `asin` grad at `x = 1` | `inf` | `inf` | `inf` |

The last two rows are there to head off the obvious wrong diagnosis: **`acos`/`asin`/`sqrt` did not change**.
Only `clamp`'s treatment of the closed boundary did.

## A reachable instance: NaN gradients from half-precision descriptor matching

`kornia/feature/matching.py:51`

```python
dm = d1_sq.repeat(1, d2.size(0)) + d2_sq.repeat(1, d1.size(0)).t() - 2.0 * d1 @ d2.t()
dm = dm.clamp(min=0.0).sqrt()
```

This is `_cdist`'s fallback branch, taken for **MPS tensors, `float16`/`bfloat16`, and under export** (the
`torch.cdist` fast path covers the rest). An exactly-zero distance — two identical descriptors — puts `dm` on the
clamp bound.

```python
import torch
from kornia.feature.matching import _cdist

for dt in (torch.float16, torch.float32):
    d1 = torch.tensor([[1.0, 0.0], [0.0, 1.0]], dtype=dt, requires_grad=True)
    d2 = torch.tensor([[1.0, 0.0], [2.0, 2.0]], dtype=dt)   # row 0 identical to d1 row 0
    dm = _cdist(d1, d2)
    dm.sum().backward()
    print(dt, dm[0, 0].item(), d1.grad.flatten().tolist())
```

```
torch 2.5.1    float16   0.0   [nan, nan, -1.6015625, 0.259765625]
torch 2.5.1    float32   0.0   [-0.44721359, -0.89442718, -1.60153389, 0.25989318]
torch 2.14.0   float16   0.0   [-0.447265625, -0.89453125, -1.6015625, 0.259765625]
torch 2.14.0   float32   0.0   [-0.44721359, -0.89442718, -1.60153389, 0.25989318]
```

`float32` on CPU takes `torch.cdist` and is finite on both. `float16` takes the clamp branch and is **`nan` on
2.5.1** and finite on 2.14. Note this is the branch with the least test coverage in the repo, and the
half-precision CI legs pin **2.9.1** — the version where it is broken.

Zero distances are not exotic here: kornia's detectors pad their output with zero LAFs, and identical padded
descriptors match at ~0 distance (this is the mechanism behind the padding self-match problem already known in
`kornia.feature`).

## Audit of the pattern

`grep -rnE "clamp[^)]*\)\.(acos|asin|atanh|sqrt|log|rsqrt)\(|(acos|asin|atanh|sqrt|log|rsqrt)\(.*clamp" kornia/`

**Bound is the singular point — affected:**

- `kornia/feature/matching.py:51` — `dm.clamp(min=0.0).sqrt()` (reproduced above)
- `kornia/geometry/solvers/polynomial_solver.py:284` — `torch.sqrt(torch.clamp(R_sq, min=0.0))`
- `kornia/geometry/solvers/polynomial_solver.py:299` — `torch.sqrt(torch.clamp(radicand, min=0.0))`
- `kornia/geometry/conversions.py:1156` — `acos(clamp(w, -1, 1))`, which is #4007 and which #4228 is fixing

**Bound is a positive floor — safe on every version, no action:**

- `kornia/feature/aliked/aliked.py:171` (`min=1e-8`), `kornia/geometry/conversions.py:473` (`min=1e-12`),
  `kornia/geometry/liegroup/se3.py:201` (`clamp_min(1e-12)`),
  `kornia/geometry/camera/distortion_kannala_brandt.py:148` (`min=1e-16`),
  `kornia/feature/loftr/utils/fine_matching.py:83` (`min=1e-10`).

I only reproduced an end-to-end NaN for `matching.py`; the two `polynomial_solver.py` sites are listed because
they match the pattern, not because I found an input that reaches them (`solve_cubic` on a degenerate cubic came
back finite on both versions).

## Suggested scope

1. **Do not fix these by clamping.** A clamp whose bound is the singularity is not a gradient guard on the
   majority of kornia's supported torch versions. The working pattern is the one #4228 uses: substitute a safe
   argument into the call that gets *differentiated*, and take the value from a `.detach()`ed copy of the real
   argument — that is version-independent by construction.
2. Fix `matching.py:51`, with a pin that differentiates through `_cdist` at an exact zero distance in
   `float16`/`bfloat16` (a `float32` CPU pin proves nothing — it takes the other branch).
3. Decide whether the two `polynomial_solver.py` sites are reachable at their bound and either fix or leave them
   with a note.
4. Worth a line in `AGENTS.md`'s precision section: `clamp` is not a portable gradient guard across kornia's
   supported torch range.

## Environment

kornia `main` @ `b56a1bc2`, torch 2.5.1 / 2.9.1 / 2.14.0, python 3.11, macOS arm64, CPU.
Related: #4007 (one instance), #4228 (fixes that instance), #4203 (the same "two ends of the supported torch
range disagree" shape, for `inv_ex`).

Posted on behalf of @ducha-aiki by Claude (Claude Opus 5).

### PR DESCRIPTION
**Pull request by mohansree14**
## What

Closes #4229 by writing its finding into `AGENTS.md`'s precision section — item 4 of that issue's suggested
scope, and the last one still open.

The other items already landed, by other contributors:

- item 2, `kornia/feature/matching.py` — #4233
- item 3, both `polynomial_solver.py` sites — #4338 and #4339
- item 1 is the principle this bullet records; #4007's instance is #4228

## The bullet

The mechanism is easy to rediscover the expensive way, because a CPU float32 run on a recent torch shows nothing:

> `torch.clamp` is not a portable gradient guard: its derivative **at the bound** passes the incoming gradient
> through (`1.0`) on torch 2.5.1 and 2.9.1 and returns `0.0` on 2.14.0. Wherever the bound is also the singular
> point — `clamp(min=0).sqrt()`, `clamp(-1, 1).acos()`, `clamp(-1, 1).asin()` — the clamp bounds the value only,
> so the unbounded derivative still reaches the backward pass as `inf` or `nan` on the older half of the
> supported range while 2.14 masks it. […] the 2.5.1 leg is what discriminates.

It also carries the half-precision corollary, which #4229 does not spell out and which I hit while auditing the
"positive floor — safe on every version" list: a floor below the dtype's smallest subnormal is not a guard
either, because it underflows to zero and the clamp becomes a no-op. `clamp(min=1e-12)` is exactly `clamp(min=0)`
in `float16` (fp16's smallest subnormal is ~6e-8).

And it names the fix that does not depend on the torch version — substitute a safe argument into the expression
that gets differentiated, take the value from a `.detach()`ed copy or the other arm of a `torch.where` — which is
what #4233, #4338, #4339 and #4228 all converged on independently.

## The grep gate

Per `AGENTS.md`, a change that closes a documented issue has to leave `grep -rnE "#4229|issues/4229" kornia/
tests/` clean of anything that reads as open work. Three hits survive:

- `kornia/enhance/jpeg.py:242` and `kornia/geometry/solvers/polynomial_solver.py:319` — both are retrospective in
  substance (they explain why the code does *not* use clamp), so following the form accepted in #4184 I reworded
  each to say it points at a measurement rather than at a tracking issue, rather than deleting a citation that is
  still useful to a reader.
- `tests/geometry/solvers/test_polynomial_solver.py:459` — a pin, which the guide says goes on naming its issue.

No `CHANGELOG.md` entry: this is contributor guidance, with no user-visible behaviour change. Happy to add one if
you would rather it were recorded there too.

## Verification

Documentation and comments only — no executable change. `ruff check` and `ruff format --check` clean on both
touched modules, both still import, and `tests/geometry/solvers/test_polynomial_solver.py -k 4229` passes (3
passed). The two `clamp` claims in the bullet are the ones measured in #4229's table; I reproduced the pre-2.14
column locally on torch 2.7.0 (`clamp(x, min=0.0)` grad at `x = 0` is `1.0`, `clamp(min=0).sqrt()` grad at `0` is
`inf`), and the `float16` underflow claim directly (`torch.tensor(0.0, dtype=torch.float16).clamp(min=1e-12)` is
`0.0`).

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Reviewed exact head `82bb1fc37746db0a333fe5a7fffbad174f91d6cd`.

[P2] Reconcile this closing PR with the documented-issue gate in AGENTS.md. The body says Closes #4229, but the prescribed issue-number grep still finds it at kornia/enhance/jpeg.py:243 and kornia/geometry/solvers/polynomial_solver.py:319. The gate explicitly says a surviving kornia/ hit makes a closing change incomplete; it does not currently exempt retrospective citations. The reworded explanations are accurate and useful, but move their measurement citation to contributor documentation or otherwise remove these source hits, or use Refs #4229 and keep the issue open. This is a repository-policy blocker, not a correctness objection to the guidance. The focused #4229 solver selection passed 3 tests; hosted checks are green.

Validation used isolated checkouts with import origins verified. No fresh whole-suite, CUDA or MPS run was performed locally.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6-astra).

**Comment by mohansree14** [comment]
Thanks — agreed, that's a policy blocker and the gate reads it correctly.

Pushed `72d4901`. Both surviving `kornia/` hits now cite the delivering PR instead of the closing issue:

- `kornia/enhance/jpeg.py:243` — `as measured in #4229` → `as measured in PR #4406`
- `kornia/geometry/solvers/polynomial_solver.py:319` — same substitution

The explanations are otherwise unchanged, and the measurement is still reachable from the source. The full audit stays in the AGENTS.md bullet (outside the `kornia/`/`tests/` grep scope), and the pin in `tests/geometry/solvers/test_polynomial_solver.py:459` keeps its `#4229` name, which the gate allows.

`grep -rnE "#4229|issues/4229" kornia/` now returns nothing, so `Closes #4229` is clean. `ruff check` and `ruff format --check` pass on both files.

**Comment by mohansree14** [comment]
@ducha-aiki this is ready for another look when you have a moment — I can't re-request review from a fork. Thanks!

**Review by ducha-aiki — APPROVED** [review]
Re-review of exact head `72d49014fee376b7a38f55fe0ec5e94913c5d475`.

The prior closing-issue blocker is fixed: both production comments now cite PR #4406, and the prescribed `#4229|issues/4229` search has no hits under `kornia/`. The remaining issue references are appropriately in contributor guidance and its regression pin. The guidance accurately captures the supported-version gradient behavior. The focused solver selection passed on CPU float32/float64 (6 cases) and MPS float32 (3 cases), and all 40 hosted checks are green. I found no blocking issues.

Validation used an isolated checkout with the import origin verified. No fresh whole-suite or CUDA run was performed locally.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).
